software/glasgow/target/toolchain.py

```diff
@@ -0,0 +1,321 @@
+        return self.name.upper().replace("-", "_").replace("+", "X")
+
+    @property
+    @abstractmethod
+    def command(self):
+        """Command name for invoking the tool.
+
+        Full path to the executable that can be used to run the tool, or ``None`` if the tool
+        is not available.
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def available(self):
+        """Tool availability.
+
+        ``True`` if the tool is installed, ``False`` otherwise. Installed binary may still not
+        be runnable, or might be too old to be useful.
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def version(self):
+        """Tool version number.
+
+        ``None`` if version number could not be determined, or a tool-specific tuple if it could.
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def identifier(self):
+        """Unique tool identifier.
+
+        Returns an array of 16 bytes that uniquely identifies the behavior of this particular tool
+        in its entirety, but has no other meaning. Typically implemented by hashing the binary and
+        its data files.
+        """
+
+    def __repr__(self):
+        return f"<{self.__class__.__module__}.{self.__class__.__name__} {self.name}>"
+
+
+
+class WasmTool(Tool):
+    PREFIX = "yowasp-"
+
+    @property
+    def python_package(self):
+        if self.name == "yosys" or self.name.startswith("nextpnr-"):
+            return self.PREFIX + self.name
+        if self.name == "icepack":
+            return self.PREFIX + "nextpnr-ice40"
+        if self.name == "ecppack":
+            return self.PREFIX + "nextpnr-ecp5"
+        raise NotImplementedError(f"Python package for tool {self.name} is not known")
+
+    @property
+    def available(self):
+        try:
+            importlib.metadata.metadata(self.python_package)
+            return True
+        except importlib.metadata.PackageNotFoundError:
+            return False
+
+    @property
+    def command(self):
+        if self.available:
+            basename = self.PREFIX + self.name
+            # We cannot assume that the command is on PATH and accessible by its basename. This
+            # will not be true when Glasgow is running from a pipx virtual environment (which isn't
+            # activated when the `glasgow` script is run). Also, our build environment does not
+            # even *have* PATH.
+            return os.path.join(sysconfig.get_path('scripts'), basename)
+
+    @property
+    def version(self):
+        if self.available:
+            # Running Wasm tools for the first time can incur a significant delay, so use
+            # the version from the Python package metadata (which is guaranteed to be the same).
+            # This makes querying the version at least as fast as for the native tools.
+            return (*importlib.metadata.version(self.python_package).split("."),)
```

Comment:
I'm intrigued by the `tuple(...)` to `(*... , )` change... no need to revert, but I'd appreciate the insight.

```suggestion
            return tuple(importlib.metadata.version(self.python_package).split("."))
```

Reply:
Less typing for the same result, plus you do not depend on the name `tuple` being defined correctly. I prefer to rely on syntax in dynlangs as much as I can. For the same reason I use e.g. list comprehensions instead of casting using `list()`, etc.

Reply:
Thanks!